google-cloud-core/src/main/java/com/google/cloud/batchingexperimental/RequestAccumulator.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.batchingexperimental;
+
+import com.google.api.core.InternalApi;
+import com.google.api.core.SettableApiFuture;
+import com.google.common.base.Preconditions;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Accumulates requests, used to batch many requests into one RPC call.
+ *
+ * <p>Experimental and only meant to be used by other packages in google-cloud-java.
+ *
+ * <p>Methods are not thread-safe.
+ *
+ * <p>Sample usage:
+ *
+ * <pre>{@code
+ * synchronized(accumulator) {
+ *   accumulator.add(item, future);
+ *   while (accumulator.hasBatch()) {
+ *     // If items and futures must be retained, they must be copied.
+ *     List<E> items = new ArrayList<>(accumulator.batch());
+ *     List<SettableApiFuture<R>> futures = new ArrayList<>(accumulator.futureBatch());
+ *
+ *     // ...
+ *
+ *     accumulator.next();
+ *   }
+ * }
+ * }</pre>
+ */
+@InternalApi
+public class RequestAccumulator<E, R> {
```

Comment:
Please document what E & R are, it's nonobvious

Reply:
Good idea. Done.

Reply:
Could you spell out E & R instead of using single-letter parameters?

Reply:
Done